### DISCUSSION TITLE
Our First Rakefile!

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#25] [Tech Stuff]
@thoni56, I've finally managed to come up with an initial Rakefile for the project, which you'll find in the [`dev-rake`][dev-rake] branch.

## Rakefile Features

Now you can use `rake` to build the whole project or a specific library only.

The `rake` command works in any repository subfolder, as if it was invoked from the root folder, so you don't need to move up and down the directory tree to issue Rake commands (which is quite cool).

To see all target tasks, use `rake -T`:

```
> rake -T
(in X:/GitHub/alan-i18n)
rake clean           # Remove any temporary products
rake clobber         # Remove any generated files
rake lib:all         # Test all libraries
rake lib:en:all      # Test English library
rake lib:en:cloak    # Cloak of Darkness
rake lib:es:all      # Test Spanish library
rake lib:es:vampiro  # Vampiro
```

As you can see, there are global tasks and namespaced tasks, organized hierarchically by domain, locale and sub-task. This should be useful when the repository grows. E.g. we'll soon add a test-suite for each library, and maybe its ADoc documentation which needs to be converted to HTML, etc., and each will be assigned its own task name (e.g. `lib:en:tests`, `lib:en:docs`, etc.).

E.g. to just rebuild the English library:

    rake lib:en:all

which will recompile the Cloak of Darkness adventure and rebuild all transcripts, but only if dependencies have changed.

To see the full dependencies tree (which includes file tasks not shown with `rake -T`) type `rake -P`:

```
> rake -P
(in X:/GitHub/alan-i18n)
rake alan_en/StartLib/StartLib.i
    phony
rake alan_en/StartLib/attack.i
    phony
[...]
rake default
    lib:all
rake lib:all
    lib:en:all
    lib:es:all
rake lib:en:all
    cloak
rake lib:en:cloak
    alan_en/cloak/cloakv3.a3c
    alan_en/cloak/cloak-stress-test.a3t
    alan_en/cloak/walkthrough.a3t
[...]
```

(the actual tasks list is too long, I've trimmed it)

To delete all generated files: `rake clobber` (only works globally, affecting all libraries).

To force build even if files didn't change: `rake -B`.

## Rake Integration Discussion

Before integrating the Rakefile into the repository, I wanted first some feedback from you, whether you like this system or not, etc.

Before merging, I'd also wait that the planned library folders renaming is in place.

IMO, once we add the Rakefile we should get rid of all the build shell scripts, since Rake now covers them all and we don't want to have to maintain double standards whenever we changes something. Also, Rake works in CMD or Bash alike, since Shell commands are invoked via Ruby's [FileUtils] library, which is OS agnostic and doesn't require an actual Bash under Windows — so, no more terminal dependency Hell, like for asciidoctor-fopub, which needed to be run using Batch scripts on Win OS, and Shell scripts on all others, now it's all Ruby code.

Of course, the `Rakefile` can (and should) still be improved to replace duplicate code with Ruby functions, and possibly add an ALAN class to handle SDK versions (switching from Alpha to Beta according to branch, using specific SDKs stored in the repository ignored folders, and even download the required SDK binaries). But these improvements will come in time.

Believe it or not, I've spent the past two weeks reading the whole Rake documentation, various tutorials and an entire book (unfortunately there isn't so much material on Rake as I'd love, but at least it's good material). It took some experimenting to get the grip on Rake, and to understand how its strange DSL syntax maps to actual Ruby code (very cleverly done), and then things started to click into place. I'm quite proud of this first breakthrough, because I believe Rake could be used on all ALAN repositories — it's so much better than Make or CMake, it's cleaner, it's simpler, and it's more powerful because it's all actual Ruby code.

I'm looking forward for your feedback on the new Rakefile!

[dev-rake]: https://github.com/alan-if/alan-i18n/tree/dev-rake
[FileUtils]: https://rubyapi.org/3.0/o/fileutils

**Comment by thoni56** [comment]
I've had my eye on rake even before you announced your experiment in this project. All attempts at creating a Make replacement is either too clumsy for my taste (e.g. Cmake) or have died (e.g. tup).

But I never really got my head around it, since I have only very shallow understanding of Ruby, but am willing to learn. Your reading list would be helpful, since, as you say, the documentation and support is very limited. I did watch Avdi Grimm's videos but for my exploration (building and running unittests) it was too shallow, and I could not find any helpers on the problem I had.

So looking forward to learning this.

However. on Ubuntu 20.04 (WSL) I got this:

```
thoni@thoni64:~/Utveckling/alan-i18n$ rake -T
rake aborted!
SyntaxError: /home/thoni/Utveckling/alan-i18n/Rakefile:31: syntax error, unexpected '=', expecting ';' or '\n'
  def devnull = "NUL"
              ^
/home/thoni/Utveckling/alan-i18n/Rakefile:33: syntax error, unexpected '=', expecting ';' or '\n'
  def devnull = "/dev/null"
              ^
/home/thoni/Utveckling/alan-i18n/Rakefile:82: dynamic constant assignment
    LIB_EN_SOURCES = FileList['alan_en/StartLib...
    ^~~~~~~~~~~~~~
/home/thoni/Utveckling/alan-i18n/Rakefile:100: dynamic constant assignment
    CLOAK_EN_A3S = FileList['alan_en/cloak/*....
    ^~~~~~~~~~~~
/home/thoni/Utveckling/alan-i18n/Rakefile:101: dynamic constant assignment
    CLOAK_EN_A3T = CLOAK_EN_A3S.ext('.a3t')
    ^~~~~~~~~~~~
/home/thoni/Utveckling/alan-i18n/Rakefile:129: dynamic constant assignment
    LIB_ES_SOURCES = FileList['alan_es/StartLib...
    ^~~~~~~~~~~~~~
/home/thoni/Utveckling/alan-i18n/Rakefile:147: dynamic constant assignment
    VAMPIRO_A3S = FileList['alan_es/vampiro/...
    ^~~~~~~~~~~
/home/thoni/Utveckling/alan-i18n/Rakefile:148: dynamic constant assignment
    VAMPIRO_A3T = VAMPIRO_A3S.ext('.a3t')
    ^~~~~~~~~~~
/usr/share/rubygems-integration/all/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
(See full trace by running task with --trace)
thoni@thoni64:~/Utveckling/alan-i18n$ ruby --version
ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]
thoni@thoni64:~/Utveckling/alan-i18n$ rake --version
rake, version 13.0.1
```

**Comment by tajmone** [comment]
I think that you're getting those errors because you're using Ruby 2.x, while I've already moved on to Ruby 3.x for all our repositories. I should add a Ruby version check to the Rakefile, to ensure users are running Ruby 3.

I don't know about Ubuntu, but on Windows you can keep multiple versions of Ruby at once and switch between them with the `ridk` command, so if you install Ruby 3 you can always revert to an older Ruby version for a given session via `ridk use 2.7`, etc.

> But I never really got my head around it, since I have only very shallow understanding of Ruby, but am willing to learn.

Mee too, I was a bit scared by Ruby's apparently mysterious syntax (all these pipes and strange symbols), but then I was lucky to find an eBooks bundle special offer for €4 that also included a 500 pages Ruby book, and within a day of reading all those strange notations started to make sense and I began to decipher the mysterious Ruby symbols — truth is, Ruby has a very friendly syntax, with all punctuation being optional whenever not strictly required (parenthesis in function/methods calls, etc.), and offers a lot of syntactic sugar. Some constructs look unfamiliar, like appending conditions to statements instead of preceding them, but when you think about it it makes a lot of sense in terms of code readability.

As for the learning resources, I also found the tutorials a bit shallow. The point is that the Rake documentation has all the required info, except examples (there are some, but not many).

I strongly advise you the video tutorial _[Goruco 2012 Power Rake]_ by Rake's creator Jim Weirich, who unfortunately passed away a couple of years after that video. That video covers advanced Rake concepts with practical examples, and there are a few questions and answers which cover common doubts too.

As for books, I've only managed to find one: _[Rake Task Management Essentials]_ by Andrey Koleshko (2014, not very recent). I didn't find it to be a particularly exciting book, but it does provide a lot of good examples to cover each topic, and it also has a _Continuous Integration_ chapter dedicated to using Rake with Jenkins, which you'll probably find quite useful. But overall, it doesn't really stand out as being a particularly in-depth book, although it does give a good panoramic view of Rake. E.g. it doesn't cover how to create `:phony` targets, which is a built-in rake extra module that is not covered much in documentation (I had to discover how it works by reading its source code).

Most of the other Rake tutorials and books are specific to Ruby on Rails, which is possibly one of Ruby's most used libraries; although these tutorials will also contain useful Rake info, they are probably not very useful unless using Rails (which extends Rake, so it might also be misleading for us).

Having said that, the Rake official documentation remains the best source of info, and I've explored all the Rake classes and methods to gain more insight (you can often click on the various methods to reveal their source code, which also helps gaining insight into their parameters meaning).



[Goruco 2012 Power Rake]: https://amara.org/en/videos/GxlygXn6h8SB/info/goruco-2012-power-rake-by-jim-weirich/
[Rake Task Management Essentials]: https://www.packtpub.com/product/rake-task-management-essentials/9781783280773

**Reply by tajmone** [reply]
@thoni56, did you manage to run the Rakefile on Ubuntu? was the problem the Ruby version?

I was hoping to merge the Rakefile into `main` branch so I can start using it, and create an update version for the Italian branch too.

Since I'll soon be introducing the test suites, having a Rakefile will reduce update times and adds the convenience of Rake being invocable from within any repository folder (no need to keep multiple Bash windows open at once).

**Reply by thoni56** [reply]
I actually had some problems getting to grips with the ruby installation. (First having to select between `RVM` and `rbenv`...) and then "debugging" the versions that Ubuntu 20.04 had.

Finally installing `rbenv` from git work fine, and from there it was just a few hickups. With a new langauge I always do some TDD so I needed to understand the unittest framework and get that to work.

Now I have

    thoni@thoni64:~/Utveckling/alan-i18n$ ruby --version
    ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]
    thoni@thoni64:~/Utveckling/alan-i18n$ rake --version
    rake, version 13.0.3

But I still got some errors:

```
thoni@thoni64:~/Utveckling/alan-i18n$ rake
cd /home/thoni/Utveckling/alan-i18n

## Compiling cloakv3.alan
#########################
cd /home/thoni/Utveckling/alan-i18n/alan_en/cloak
alan -include /home/thoni/Utveckling/alan-i18n/alan_enStartLib cloakv3.alan 1>/dev/null
rake aborted!
Command failed with status (1): [alan -include /home/thoni/Utveckling/alan-...]
/home/thoni/Utveckling/alan-i18n/Rakefile:56:in `block in <top (required)>'
Tasks: TOP => default => lib:all => lib:en:all => lib:en:cloak => alan_en/cloak/cloakv3.a3c
(See full trace by running task with --trace)
```
(The trace doesn't add much...)

I'm suspicious about the `.../alan-i18n/alan_enStartLib` in the alan command. Shouldn't that be `.../alan-i18n/alan_en/StartLib`, with a slash in the last path?

I'm on Ubuntu 20.04 in WSL.

**Reply by tajmone** [reply]
> Now I have
> 
> ```
> thoni@thoni64:~/Utveckling/alan-i18n$ ruby --version
> ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]
> thoni@thoni64:~/Utveckling/alan-i18n$ rake --version
> rake, version 13.0.3
> ```

Should be fine, on Windows I have same Ruby version, but Rake 13.0.6, so small difference in the batch version (since under Win Rake is built into Ruby, I had to install it manually as a gem to get latest version, overriding the default Rake which is a binary executable).

> ```
> ## Compiling cloakv3.alan
> #########################
> cd /home/thoni/Utveckling/alan-i18n/alan_en/cloak
> alan -include /home/thoni/Utveckling/alan-i18n/alan_enStartLib cloakv3.alan 1>/dev/null
> ```

I see the problem, it's not inserting the path separator:

    alan-i18n/alan_enStartLib cloakv3.alan

should really be:

    alan-i18n/alan_en/StartLib cloakv3.alan

the separator is actually handled by Ruby code I've added, which I hoped would insert the correct native dir separator according to OS. In the [Rakefile]:

```ruby
  lib_dir = File::expand_path(t.source.pathmap("%1d") << "#{File::ALT_SEPARATOR}Foundation")
```

So I need to look again into the `File::ALT_SEPARATOR`, which I understood would always emit the OS specific dir separator, but apparently this only works on Windows. Should be an easy fix.




[Rakefile]: https://github.com/alan-if/alan-i18n/blob/dev-rake/Rakefile#L47

**Reply by tajmone** [reply]
Note that I've forced pushed the Rake branch, since I've now renamed the StartLib folder to Foundation,, etc. But the problem is definitely there, regardless of the branch update.

**Reply by tajmone** [reply]
## Fixed!

I've updated the Rakefile, now it should work on Linux too.

**Reply by thoni56** [reply]
Works!

**Reply by tajmone** [reply]
Great! I then suggest to merge it into `main` branch as soon as I've finished polishing it here and there and update the docs about it (e.g. mention in each project the Rake command that can trigger that specific build (e.g. a specific locale library, or just for the _Cloak of Darkness_ or _Vampiro_ adventures, and so on).

Should we then just drop the shell scripts? It should be safe to assume that contributors to this project will have Ruby installed, especially since the library and project docs will eventually be in AsciiDoc — at least, a safe enough assumption to justify doing away with maintenance redundancy.

**Comment by thoni56** [comment]
Honestly, I haven't gotten around to using the shell scripts in `i8n`, so removing them in favour of Rakefiles is fine by me.
